File: Leetcode/剑指offer/面试题07. 重建二叉树.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
var buildTree = function(preorder, inorder) {
  if(preorder.length === 0 || inorder.length === 0 || (preorder.length !== inorder.length)) {
    return null;
  }
  let rootVal = preorder[0];
  let index = inorder.indexOf(rootVal);
  let root = new TreeNode(rootVal);
  root.left = buildTree(preorder.slice(1, index+1), inorder.slice(0, index));
  root.right = buildTree(preorder.slice(index+1), inorder.slice(index+1))
  return root;
};

File: Leetcode/剑指offer/面试题07.重建二叉树.py
'''
https://leetcode-cn.com/problems/zhong-jian-er-cha-shu-lcof/
已知二叉树前中序遍历，重构二叉树。python3-递归法
'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        if not preorder or not inorder:
            return False
        root = TreeNode(preorder[0])

        inleft = inright = preleft = preright =  []
        tree_len = len(inorder)
        for node in range(tree_len):
            if inorder[node] == preorder[0]:
                inleft = inorder[0:node]
                inright = inorder[node+1:tree_len]
                left_len = len(inleft)
                preleft = preorder[1:left_len+1]
                preright = preorder[left_len+1:]
        
        if inleft:
            root.left = self.buildTree(preleft,inleft)
        if inright:
            root.right = self.buildTree(preright,inright)
        return root
